src/main/java/com/google/devtools/build/lib/syntax/MethodDescriptor.java

```diff
@@ -59,19 +62,157 @@ private MethodDescriptor(
+        mh = MethodHandles.collectArguments(fromJava, 0, mh);
+      }
+
+      mh = MethodHandleUtil.moveArgument(mh, 3, 2);
+
+      Class<?> expectedP0 = isStringModule ? String.class : method.getDeclaringClass();
+      // Self-check: better fail earlier.
+      Preconditions.checkState(mh.type().parameterCount() == 4, "%s", mh.type());
+      Preconditions.checkState(mh.type().parameterType(0) == expectedP0, "%s", mh.type());
+      Preconditions.checkState(mh.type().parameterType(1) == StarlarkThread.class, "%s", mh.type());
+      Preconditions.checkState(mh.type().parameterType(2) == Mutability.class, "%s", mh.type());
+      Preconditions.checkState(mh.type().parameterType(3) == Object[].class, "%s", mh.type());
+
+      mh =
+          MethodHandles.explicitCastArguments(
+              mh,
+              MethodType.methodType(
+                  Object.class,
+                  Object.class,
+                  StarlarkThread.class,
+                  Mutability.class,
+                  Object[].class));
+
+      this.methodHandle = mh;
+    } catch (Exception e) {
+      throw new IllegalStateException(e);
+    }
+  }
+
+  private static MethodHandle handleNullReturn(
+      MethodHandle mh, boolean allowReturnNones, String name) {
+    if (mh.type().returnType().isPrimitive()) {
+      return mh;
+    }
+
+    if (allowReturnNones) {
+      // TODO(adonovan): eliminate allowReturnNones. Given that we convert
+      //   String/Integer/Boolean/List/Map, it seems obtuse to crash instead
+      //   of converting null too.
+      return nullToNone(mh);
+    } else {
+      return nullToException(mh, name);
+    }
+  }
+
+  private static Object nullToNone(Object object) {
+    return object != null ? object : Starlark.NONE;
+  }
+
+  private static Object nullToException(Object object, String methodName) {
+    if (object == null) {
+      throw new IllegalStateException("method invocation returned None: " + methodName);
+    }
+    return object;
+  }
+
+  private static MethodHandle nullToNone(MethodHandle mh) {
+    MethodHandle filter =
+        MethodHandles.explicitCastArguments(
+            NULL_TO_NONE, MethodType.methodType(Object.class, mh.type().returnType()));
+    return MethodHandles.filterReturnValue(mh, filter);
+  }
+
+  private static MethodHandle nullToException(MethodHandle mh, String name) {
+    MethodHandle filter = MethodHandles.insertArguments(NULL_TO_EXCEPTION, 1, name);
+    filter =
+        MethodHandles.explicitCastArguments(
+            filter, MethodType.methodType(mh.type().returnType(), mh.type().returnType()));
+    return MethodHandles.filterReturnValue(mh, filter);
+  }
+
+  // Cache method handles to avoid looking them up again and again
+  private static final MethodHandle NULL_TO_NONE;
+  private static final MethodHandle NULL_TO_EXCEPTION;
+  private static final MethodHandle FROM_JAVA;
+
+  static {
+    try {
+      NULL_TO_NONE =
+          MethodHandles.lookup()
```

Comment:
It's too bad there's no way to say something like NULL_TO_NONE = MethodHandle.of(MethodDescriptor::nullToNone) and avoid the redundancy and indirection. Sadly the :: syntax triggers unwanted lambda creation.

Reply:
Good news is that Idea completes `lookup.find*` arguments, and highlights it as a warning if parameter types are known and incorrect.

> indirection

Did not get where is indirection here.